File: public/project/recommendations.py
"""
recommendations.py
==================
Contributing-factor analysis, personalized recommendations and a 7-day action
plan for the Smart Academic Stress Analyzer.

IMPORTANT
---------
The actual Stress / Burnout *scores* are produced by ``fuzzy_system.py``
using Scikit-Fuzzy (Mamdani + centroid defuzzification). This module only
explains the *contributing factors* and suggests *what to do* about them,
using simple if/else heuristics on the raw student inputs (which is allowed).
"""

from fuzzy_system import categorize


# Thresholds used for factor detection (kept consistent with the project).
SLEEP_LOW = 6.0        # hours
STUDY_HIGH = 10.0      # hours
WORKLOAD_HIGH = 6.0    # 0-10 scale
ATTENDANCE_LOW = 70.0  # percent


def contributing_factors(inputs):
    """
    Return a list of contributing factors based on the raw student inputs.

    Each factor: {factor, value, unit, detail, severity}
    """
    factors = []

    if inputs['sleep'] < SLEEP_LOW:
        gap = round(SLEEP_LOW - inputs['sleep'], 1)
        factors.append({
            'factor': 'Low Sleep',
            'value': inputs['sleep'],
            'unit': 'hours',
            'severity': 'High' if inputs['sleep'] < 4 else 'Medium',
            'detail': f"Sleeping {inputs['sleep']}h is below the healthy "
                      f"minimum (~{SLEEP_LOW:.0f}h); a deficit of {gap}h.",
        })

    if inputs['study'] > STUDY_HIGH:
        factors.append({
            'factor': 'High Study Hours',
            'value': inputs['study'],
            'unit': 'hours',
            'severity': 'High' if inputs['study'] >= 12 else 'Medium',
            'detail': f"Studying {inputs['study']}h/day is well above the "
                      f"recommended focus budget (~{STUDY_HIGH:.0f}h).",
        })

    if inputs['assignment'] > WORKLOAD_HIGH:
        factors.append({
            'factor': 'High Assignment Workload',
            'value': inputs['assignment'],
            'unit': '/10',
            'severity': 'High' if inputs['assignment'] >= 8 else 'Medium',
            'detail': f"Workload rated {inputs['assignment']}/10 indicates a "
                      f"heavy pile of pending assignments.",
        })

    if inputs['attendance'] < ATTENDANCE_LOW:
        factors.append({
            'factor': 'Low Attendance',
            'value': inputs['attendance'],
            'unit': '%',
            'severity': 'High' if inputs['attendance'] < 50 else 'Medium',
            'detail': f"Attendance at {inputs['attendance']}% is below the "
                      f"{ATTENDANCE_LOW:.0f}% expected for steady progress.",
        })

    if not factors:
        factors.append({
            'factor': 'Balanced Profile',
            'value': None,
            'unit': '',
            'severity': 'Low',
            'detail': 'All inputs are within healthy ranges. Keep it up!',
        })

    return factors


def recommendations(inputs, stress_level, burnout_level):
    """
    Personalized recommendations driven by the raw inputs (simple if/else).
    """
    recs = []

    if inputs['sleep'] < SLEEP_LOW:
        recs.append({
            'title': 'Improve your sleep routine',
            'detail': 'Aim for 7-8 hours. Set a fixed bedtime, avoid screens '
                      '30 min before sleep, and cut caffeine after 4 PM.',
        })

    if inputs['study'] > STUDY_HIGH:
        recs.append({
            'title': 'Use shorter, focused study sessions',
            'detail': 'Switch to Pomodoro (25 min focus + 5 min break). '
                      'Spacing study blocks improves retention and lowers fatigue.',
        })

    if inputs['assignment'] > WORKLOAD_HIGH:
        recs.append({
            'title': 'Break assignments into smaller tasks',
            'detail': 'Split each assignment into sub-tasks, prioritise by '
                      'deadline, and tackle one piece at a time to avoid overload.',
        })

    if inputs['attendance'] < ATTENDANCE_LOW:
        recs.append({
            'title': 'Review missed classes & improve attendance',
            'detail': 'Catch up on missed lectures using notes/recordings and '
                      'commit to attending upcoming classes to stay on track.',
        })

    if stress_level == 'High' or burnout_level == 'High':
        recs.append({
            'title': 'Reach out for support',
            'detail': 'High stress/burnout levels were detected. Consider '
                      'talking to a mentor, counsellor, or campus support service.',
        })

    if not recs:
        recs.append({
            'title': 'Maintain your healthy routine',
            'detail': 'Your profile looks balanced. Keep monitoring your sleep '
                      'and workload to stay in the green zone.',
        })

    return recs


def action_plan(inputs, stress_level, burnout_level):
    """
    A simple, personalised 7-day action plan derived from the inputs.
    """
    low_sleep = inputs['sleep'] < SLEEP_LOW
    high_study = inputs['study'] > STUDY_HIGH
    high_load = inputs['assignment'] > WORKLOAD_HIGH
    low_att = inputs['attendance'] < ATTENDANCE_LOW

    plan = []

    # Day 1 — foundation
    tasks = ['Audit your current weekly schedule (study, sleep, classes).']
    if low_sleep:
        tasks.append('Set a fixed bedtime tonight targeting 7-8h of sleep.')
    if high_load:
        tasks.append('List every pending assignment and rank by deadline.')
    plan.append({'day': 1, 'focus': 'Audit & plan', 'tasks': tasks})

    # Day 2 — sleep
    tasks = ['Begin a wind-down routine 45 min before bed (no screens).']
    if low_sleep:
        tasks.append('Go to bed 30 min earlier than usual.')
    plan.append({'day': 2, 'focus': 'Sleep recovery', 'tasks': tasks})

    # Day 3 — study technique
    tasks = ['Try two Pomodoro focus blocks (25 min + 5 min break).']
    if high_study:
        tasks.append('Cap total study at 6 focused hours; rest the remainder.')
    plan.append({'day': 3, 'focus': 'Smarter study', 'tasks': tasks})

    # Day 4 — assignments
    tasks = ['Pick one assignment and break it into 3-5 sub-tasks.']
    if high_load:
        tasks.append('Complete the first sub-task of your hardest assignment.')
    plan.append({'day': 4, 'focus': 'Tame the workload', 'tasks': tasks})

    # Day 5 — attendance
    tasks = ['Attend all scheduled classes today.']
    if low_att:
        tasks.append('Collect notes/recordings for 2 missed lectures and review them.')
    plan.append({'day': 5, 'focus': 'Attendance catch-up', 'tasks': tasks})

    # Day 6 — active rest
    tasks = ['Take a 20-30 min walk or light exercise.']
    if stress_level in ('High', 'Medium') or burnout_level in ('High', 'Medium'):
        tasks.append('Do a 5-minute breathing/relaxation exercise.')
    plan.append({'day': 6, 'focus': 'Active rest', 'tasks': tasks})

    # Day 7 — review
    tasks = ['Re-assess your stress using this analyzer with updated inputs.']
    tasks.append('Keep what worked this week; adjust one habit that did not.')
    plan.append({'day': 7, 'focus': 'Review & adjust', 'tasks': tasks})

    return plan


if __name__ == '__main__':
    demo = {'sleep': 4, 'study': 12, 'assignment': 8, 'attendance': 55}
    print('Factors:', contributing_factors(demo))
    print('Recs   :', recommendations(demo, 'High', 'High'))
    print('Plan   :', action_plan(demo, 'High', 'High'))

File: src/lib/recommendations.ts
/**
 * recommendations.ts — mirrors public/project/recommendations.py.
 * Contributing factors + personalized recommendations + 7-day action plan.
 *
 * NOTE: only the *advice* uses simple if/else on the raw inputs. The actual
 * Stress/Burnout scores always come from the Scikit-Fuzzy engine.
 */
import { INPUTS, type PredictionResult } from './fuzzyEngine';

const SLEEP_LOW = 6;
const STUDY_HIGH = 10;
const WORKLOAD_HIGH = 6;
const ATTENDANCE_LOW = 70;

type Severity = 'Low' | 'Medium' | 'High';

export interface Factor {
  factor: string;
  value: number | null;
  unit: string;
  severity: Severity;
  detail: string;
}

export interface Recommendation {
  title: string;
  detail: string;
}

export interface DayPlan {
  day: number;
  focus: string;
  tasks: string[];
}

export function contributingFactors(inputs: Record<string, number>): Factor[] {
  const factors: Factor[] = [];

  if (inputs.sleep < SLEEP_LOW) {
    const gap = Math.round((SLEEP_LOW - inputs.sleep) * 10) / 10;
    factors.push({
      factor: 'Low Sleep',
      value: inputs.sleep,
      unit: 'hrs',
      severity: inputs.sleep < 4 ? 'High' : 'Medium',
      detail: `Sleeping ${inputs.sleep}h is below the healthy minimum (~${SLEEP_LOW}h); a deficit of ${gap}h.`,
    });
  }
  if (inputs.study > STUDY_HIGH) {
    factors.push({
      factor: 'High Study Hours',
      value: inputs.study,
      unit: 'hrs',
      severity: inputs.study >= 12 ? 'High' : 'Medium',
      detail: `Studying ${inputs.study}h/day is well above the recommended focus budget (~${STUDY_HIGH}h).`,
    });
  }
  if (inputs.assignment > WORKLOAD_HIGH) {
    factors.push({
      factor: 'High Assignment Workload',
      value: inputs.assignment,
      unit: '/10',
      severity: inputs.assignment >= 8 ? 'High' : 'Medium',
      detail: `Workload rated ${inputs.assignment}/10 indicates a heavy pile of pending assignments.`,
    });
  }
  if (inputs.attendance < ATTENDANCE_LOW) {
    factors.push({
      factor: 'Low Attendance',
      value: inputs.attendance,
      unit: '%',
      severity: inputs.attendance < 50 ? 'High' : 'Medium',
      detail: `Attendance at ${inputs.attendance}% is below the ${ATTENDANCE_LOW}% expected for steady progress.`,
    });
  }
  if (factors.length === 0) {
    factors.push({
      factor: 'Balanced Profile',
      value: null,
      unit: '',
      severity: 'Low',
      detail: 'All inputs are within healthy ranges. Keep it up!',
    });
  }
  return factors;
}

export function recommendations(
  inputs: Record<string, number>,
  stressLevel: string,
  burnoutLevel: string
): Recommendation[] {
  const recs: Recommendation[] = [];

  if (inputs.sleep < SLEEP_LOW) {
    recs.push({
      title: 'Improve your sleep routine',
      detail: 'Aim for 7-8 hours. Set a fixed bedtime, avoid screens 30 min before sleep, and cut caffeine after 4 PM.',
    });
  }
  if (inputs.study > STUDY_HIGH) {
    recs.push({
      title: 'Use shorter, focused study sessions',
      detail: 'Switch to Pomodoro (25 min focus + 5 min break). Spacing study blocks improves retention and lowers fatigue.',
    });
  }
  if (inputs.assignment > WORKLOAD_HIGH) {
    recs.push({
      title: 'Break assignments into smaller tasks',
      detail: 'Split each assignment into sub-tasks, prioritise by deadline, and tackle one piece at a time to avoid overload.',
    });
  }
  if (inputs.attendance < ATTENDANCE_LOW) {
    recs.push({
      title: 'Review missed classes & improve attendance',
      detail: 'Catch up on missed lectures using notes/recordings and commit to attending upcoming classes to stay on track.',
    });
  }
  if (stressLevel === 'High' || burnoutLevel === 'High') {
    recs.push({
      title: 'Reach out for support',
      detail: 'High stress/burnout levels were detected. Consider talking to a mentor, counsellor, or campus support service.',
    });
  }
  if (recs.length === 0) {
    recs.push({
      title: 'Maintain your healthy routine',
      detail: 'Your profile looks balanced. Keep monitoring your sleep and workload to stay in the green zone.',
    });
  }
  return recs;
}

export function actionPlan(
  inputs: Record<string, number>,
  stressLevel: string,
  burnoutLevel: string
): DayPlan[] {
  const lowSleep = inputs.sleep < SLEEP_LOW;
  const highStudy = inputs.study > STUDY_HIGH;
  const highLoad = inputs.assignment > WORKLOAD_HIGH;
  const lowAtt = inputs.attendance < ATTENDANCE_LOW;

  const plan: DayPlan[] = [];

  const d1: string[] = ['Audit your current weekly schedule (study, sleep, classes).'];
  if (lowSleep) d1.push('Set a fixed bedtime tonight targeting 7-8h of sleep.');
  if (highLoad) d1.push('List every pending assignment and rank by deadline.');
  plan.push({ day: 1, focus: 'Audit & plan', tasks: d1 });

  const d2: string[] = ['Begin a wind-down routine 45 min before bed (no screens).'];
  if (lowSleep) d2.push('Go to bed 30 min earlier than usual.');
  plan.push({ day: 2, focus: 'Sleep recovery', tasks: d2 });

  const d3: string[] = ['Try two Pomodoro focus blocks (25 min + 5 min break).'];
  if (highStudy) d3.push('Cap total study at 6 focused hours; rest the remainder.');
  plan.push({ day: 3, focus: 'Smarter study', tasks: d3 });

  const d4: string[] = ['Pick one assignment and break it into 3-5 sub-tasks.'];
  if (highLoad) d4.push('Complete the first sub-task of your hardest assignment.');
  plan.push({ day: 4, focus: 'Tame the workload', tasks: d4 });

  const d5: string[] = ['Attend all scheduled classes today.'];
  if (lowAtt) d5.push('Collect notes/recordings for 2 missed lectures and review them.');
  plan.push({ day: 5, focus: 'Attendance catch-up', tasks: d5 });

  const d6: string[] = ['Take a 20-30 min walk or light exercise.'];
  if (stressLevel === 'High' || stressLevel === 'Medium' || burnoutLevel === 'High' || burnoutLevel === 'Medium') {
    d6.push('Do a 5-minute breathing/relaxation exercise.');
  }
  plan.push({ day: 6, focus: 'Active rest', tasks: d6 });

  const d7: string[] = [
    'Re-assess your stress using this analyzer with updated inputs.',
    'Keep what worked this week; adjust one habit that did not.',
  ];
  plan.push({ day: 7, focus: 'Review & adjust', tasks: d7 });

  return plan;
}

export const FACTOR_THRESHOLDS = { SLEEP_LOW, STUDY_HIGH, WORKLOAD_HIGH, ATTENDANCE_LOW };

export function inputsList() {
  return INPUTS;
}

export type { PredictionResult };
